DirectProgramming/DPC++FPGA/ReferenceDesigns/merge_sort/src/merge_sort.cpp

```diff
@@ -0,0 +1,334 @@
+#include <algorithm>
+#include <array>
+#include <chrono>
+#include <limits>
+#include <numeric>
+#include <type_traits>
+#include <vector>
+#include <utility> 
+
+#include <CL/sycl.hpp>
+#include <CL/sycl/INTEL/fpga_extensions.hpp>
+
+// dpc_common.hpp can be found in the dev-utilities include folder.
+// e.g., $ONEAPI_ROOT/dev-utilities/include/dpc_common.hpp
+#include "dpc_common.hpp"
+
+#include "MergeSort.hpp"
+
+using namespace sycl;
+using namespace std::chrono;
+
+// determines whether we will use USM host or device allocations to move data
+// between host and the device.
+#if defined(USM_HOST_ALLOCATIONS)
+constexpr bool kUseUSMHostAllocation = true;
+#else
+constexpr bool kUseUSMHostAllocation = false;
+#endif
+
+// the number of merge units, must be a power of 2
+#ifndef MERGE_UNITS
+#ifdef FPGA_EMULATOR
```

Comment:
Is it common practice to modify behaviour inside the source code based on whether we are running emulation or not?  I'm not a big fan of this practice, I'd rather see changes like that happen in Makefiles / cmake files.  The danger is you don't realize you're not emulating the code you think you are.

What happens if we emulate 16 merge units? If the run takes too long, wouldn't it make sense to just sort fewer things in emulation, rather than changing the structure of the code?

I'm open to being convinced that the way you've done it is ok.

Reply:
Fixed. This was a biproduct of my debugging.